File: app/src/main/java/com/fentury/testapp/activity/MainActivity.java
package com.fentury.testapp.activity;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.fentury.testapp.R;
import com.fentury.testapp.adapter.TopStoriesAdapter;
import com.fentury.testapp.api.StoriesApi;
import com.fentury.testapp.model.Model;
import com.fentury.testapp.utils.Constants;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity {
    @BindView(R.id.card_recycler_view)
    RecyclerView recyclerView;
    @BindView(R.id.empty_view)
    TextView emptyView;
    private TopStoriesAdapter topStoriesAdapter;
    private List<Integer> topStories = new ArrayList<>();
    private Retrofit retrofit;
    private StoriesApi request;
    private List<Model> models = new ArrayList<>();
    private int mLoadPosition = -1;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        initViews();
        initRetrofit();
        loadJSON();
    }

    private void initRetrofit() {
        retrofit = new Retrofit.Builder()
                .baseUrl(Constants.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        request = retrofit.create(StoriesApi.class);
    }

    private void initViews() {
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        topStoriesAdapter = new TopStoriesAdapter(models, this);
        recyclerView.setAdapter(topStoriesAdapter);
        if (topStories.isEmpty()) {
            recyclerView.setVisibility(View.VISIBLE);
            emptyView.setVisibility(View.GONE);
        } else {
            recyclerView.setVisibility(View.GONE);
            emptyView.setVisibility(View.VISIBLE);
        }
    }

    private void loadJSON() {
        Call<List<Integer>> call = request.getTopStories();
        call.enqueue(new Callback<List<Integer>>() {
            @Override
            public void onResponse(Call<List<Integer>> call, Response<List<Integer>> response) {
                topStories = response.body();
                loadDetails();
            }

            @Override
            public void onFailure(Call<List<Integer>> call, Throwable t) {
                Toast.makeText(MainActivity.this, R.string.error, Toast.LENGTH_LONG).show();
            }
        });
    }

    private void loadDetails() {
        mLoadPosition++;
        if (mLoadPosition == topStories.size()) {
            return;
        }
        int id = topStories.get(mLoadPosition);
        request.getTopStore(id).enqueue(new Callback<Model>() {
            @Override
            public void onResponse(Call<Model> call, Response<Model> response) {
                models.add(response.body());
                topStoriesAdapter.notifyDataSetChanged();
                loadDetails();
            }

            @Override
            public void onFailure(Call<Model> call, Throwable t) {
                Toast.makeText(MainActivity.this, R.string.error, Toast.LENGTH_LONG).show();
            }
        });
    }
}

File: app/src/main/java/com/fentury/testapp/activity/MainActivity.kt
package com.fentury.testapp.activity

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.TextView
import android.widget.Toast
import com.fentury.testapp.R
import com.fentury.testapp.adapter.TopStoriesAdapter
import com.fentury.testapp.api.StoriesApi
import com.fentury.testapp.model.Model
import com.fentury.testapp.utils.Constants
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.*

class MainActivity : AppCompatActivity() {
    var recyclerView: RecyclerView? = null
    var emptyView: TextView? = null
    private var topStoriesAdapter: TopStoriesAdapter? = null
    private var topStories: List<Int> = ArrayList()
    private var retrofit: Retrofit? = null
    private var request: StoriesApi? = null
    private val models = ArrayList<Model>()
    private var mLoadPosition = -1


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initViews()
        initRetrofit()
        loadJSON()
    }

    private fun initRetrofit() {
        retrofit = Retrofit.Builder()
                .baseUrl(Constants.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        request = retrofit!!.create(StoriesApi::class.java)
    }

    private fun initViews() {
        recyclerView = findViewById(R.id.card_recycler_view) as RecyclerView
        emptyView = findViewById(R.id.empty_view) as TextView

        recyclerView!!.layoutManager = LinearLayoutManager(this)
        topStoriesAdapter = TopStoriesAdapter(models, this)
        recyclerView!!.adapter = topStoriesAdapter
        if (topStories.isEmpty()) {
            recyclerView!!.visibility = View.VISIBLE
            emptyView!!.visibility = View.GONE
        } else {
            recyclerView!!.visibility = View.GONE
            emptyView!!.visibility = View.VISIBLE
        }
    }

    private fun loadJSON() {
        val call = request!!.getTopStories()
        call.enqueue(object : Callback<List<Int>> {
            override fun onResponse(call: Call<List<Int>>, response: Response<List<Int>>) {
                topStories = response.body()
                loadDetails()
            }

            override fun onFailure(call: Call<List<Int>>, t: Throwable) {
                Toast.makeText(this@MainActivity, R.string.error, Toast.LENGTH_LONG).show()
            }
        })
    }

    private fun loadDetails() {
        mLoadPosition++
        if (mLoadPosition == topStories.size) {
            return
        }
        val id = topStories[mLoadPosition]
        request!!.getTopStore(id).enqueue(object : Callback<Model> {
            override fun onResponse(call: Call<Model>, response: Response<Model>) {
                models.add(response.body())
                topStoriesAdapter!!.notifyDataSetChanged()
                loadDetails()
            }

            override fun onFailure(call: Call<Model>, t: Throwable) {
                Toast.makeText(this@MainActivity, R.string.error, Toast.LENGTH_LONG).show()
            }
        })
    }
}
